### PR TITLE
Fix re-sync from bad position

### PR DESCRIPTION
**Pull request by ianzapolsky**
This pull request fixes an issue where siddontang skips a transaction when re-connecting to the database after its connection is severed.

### CODE REVIEW

**Comment by siddontang** [comment]
Thanks @ianzapolsky 

CI failed.

File: replication/binlogsyncer.go [hunk 1]
```diff
@@ -730,26 +731,44 @@ func (b *BinlogSyncer) parseEvent(s *BinlogStreamer, data []byte) error {
...
-		if b.gset == nil {
+		if b.prevGset == nil {
 			break
 		}
 		u, _ := uuid.FromBytes(event.SID)
-		err := b.gset.Update(fmt.Sprintf("%s:%d", u.String(), event.GNO))
-		if err != nil {
-			return errors.Trace(err)
+		if b.currGset == nil {
+			gset, err := ParseGTIDSet(MySQLFlavor, u.String())
```
Comment (GregoryIan) [inline replication/binlogsyncer.go]: here `b.currGset` is a single gtid, not a full gtid set, right?

File: replication/binlogsyncer.go [hunk 2]
```diff
@@ -730,26 +731,44 @@ func (b *BinlogSyncer) parseEvent(s *BinlogStreamer, data []byte) error {
...
+			}
+			b.currGset = gset
+		} else {
+			err := b.currGset.Update(fmt.Sprintf("%d-%d-%d", GTID.DomainID, GTID.ServerID, GTID.SequenceNumber))
+			if err != nil {
+				return errors.Trace(err)
+			}
 		}
 	case *XIDEvent:
+		b.prevGset = b.currGset.Clone()
```
Comment (GregoryIan) [inline replication/binlogsyncer.go]: as comment ditto, `b.currGset` would miss some gtid, right?

File: replication/binlogsyncer.go [hunk 3]
```diff
@@ -730,26 +731,44 @@ func (b *BinlogSyncer) parseEvent(s *BinlogStreamer, data []byte) error {
...
 		event.GSet = b.getGtidSet()
 	case *QueryEvent:
+		b.prevGset = b.currGset.Clone()
```
Comment (amyangfei) [inline replication/binlogsyncer.go]: `b.currGset` may be nil, is it better to add a function like `getGitdSet`

**Comment by GregoryIan** [comment]
ping @ianzapolsky , thanks for your pr, please fix CI

**Comment by GregoryIan** [comment]
BTW, please also fix the issue 

does this still have a problem with the `Query` event BEGIN in DML event sequences like following
```
| bin.000001 | 16597 | Gtid           |       101 |       16662 | SET @@SESSION.GTID_NEXT= '9afe121c-40c2-11e9-9ec7-0242ac110002:55'                                                                                                                                                                                                                                                                                                                                        |
| bin.000001 | 16662 | Query          |       101 |       16730 | BEGIN                                                                                                                                                                                                                                                                                                                                                                                                     |
| bin.000001 | 16730 | Table_map      |       101 |       16791 | table_id: 11993 (sharding1.t2)                                                                                                                                                                                                                                                                                                                                                                            |
| bin.000001 | 16791 | Update_rows    |       101 |       17277 | table_id: 11993 flags: STMT_END_F                                                                                                                                                                                                                                                                                                                                                                         |
| bin.000001 | 17277 | Xid            |       101 |       17308 | COMMIT /* xid=51083 */
```

_Originally posted by @amyangfei in https://github.com/siddontang/go-mysql/pull/416_

**Comment by ianzapolsky** [comment]
@gleonid has a good solution to this in #420. Let's go with that.

**Comment by siddontang** [comment]
thanks @ianzapolsky 

I will close this PR.

**Pull request closed without merging**